src/handle_list.rs

```diff
@@ -0,0 +1,227 @@
+        loop {
+            // Safety
+            // This is save, because we have not stored the Ptr elsewhere so we have exclusive
+            // access.
+            // The Ptr is also still valid, because we never free Entries on the List
+            unsafe { (*n_node_ptr).next = current_head };
+
+            // Attempt to add the Entry to the List by setting it as the new Head
+            match self.inner.head.compare_exchange(
+                current_head,
+                n_node_ptr,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => return,
+                Err(n_head) => {
+                    // Store the found Head-Ptr to avoid an extra load at the start of every loop
+                    current_head = n_head;
+                }
+            }
+        }
+    }
+
+    /// Creates a new Snapshot of the List at this Point in Time
+    pub fn snapshot(&self) -> ListSnapshot {
+        ListSnapshot {
+            head: self.inner.head.load(Ordering::SeqCst),
+            _list: self.inner.clone(),
+        }
+    }
+
+    /// Inserts the Items of the Iterator, but in reverse order
+    #[cfg(test)]
+    pub fn extend<I>(&self, iter: I)
+    where
+        I: IntoIterator<Item = Arc<AtomicUsize>>,
+    {
+        for item in iter.into_iter() {
+            self.add_counter(item);
+        }
+    }
+}
+
+impl Default for HandleList {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+impl Debug for HandleList {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        // TODO
+        // Figure out how exactly we want the Debug output to look
+        write!(f, "HandleList")
+    }
+}
+impl Clone for HandleList {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+}
+
+impl ListSnapshot {
+    /// Obtain an iterator over the Entries in this Snapshot
+    pub fn iter(&self) -> SnapshotIter {
+        SnapshotIter { current: self.head }
+    }
+}
+
+impl Iterator for SnapshotIter {
+    // TODO
+    // Maybe don't return an owned Value here
+    type Item = Arc<AtomicUsize>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current.is_null() {
+            return None;
+        }
+
+        // Safety
+        // The Ptr is not null, because of the previous if-statement.
+        // The Data is also not freed, because we never free Entries on the List.
+        // We also have no one mutating Entries on the List and therefore we can access this without
+        // any extra synchronization needed.
+        let entry = unsafe { &*self.current };
+
+        self.current = entry.next;
+
+        Some(entry.data.clone())
+    }
+}
+
+impl Drop for InnerList {
+    fn drop(&mut self) {
+        // We iterate over all the Entries of the List and free every Entry of the List
+        let mut current = self.head.load(Ordering::SeqCst);
```

Comment:
I think you can use `self.head.get_mut()` here since we have `&mut self`.

Reply:
Not really, because we just need the actual Pointer stored in the head. So for now I still replaced it with `*self.head.get_mut()` but I'm not really sure if this actually better than the load.